ffi/Gosu_Song.cpp

```diff
@@ -1,52 +1,51 @@
-#include <Gosu/Audio.hpp>
-#include "Gosu_Song.h"
-
-extern "C" {
-  Gosu_Song* Gosu_Song_create(const char* filename)
-  {
-    return reinterpret_cast<Gosu_Song*>( new Gosu::Song(filename) );
-  }
-
-  void Gosu_Song_destroy(Gosu_Song* song)
-  {
-    delete( reinterpret_cast<Gosu::Song*>( song ));
-  }
-
-  void Gosu_Song_play(Gosu_Song* song, bool looping)
-  {
-    reinterpret_cast<Gosu::Song*>( song )->play(looping);
-  }
-
-  bool Gosu_Song_playing(Gosu_Song* song)
-  {
-    return reinterpret_cast<Gosu::Song*>( song )->playing();
-  }
-
-  void Gosu_Song_pause(Gosu_Song* song)
-  {
-    reinterpret_cast<Gosu::Song*>( song )->pause();
-  }
-
-  bool Gosu_Song_paused(Gosu_Song* song)
-  {
-    return reinterpret_cast<Gosu::Song*>( song )->paused();
-  }
-
-  void Gosu_Song_stop(Gosu_Song* song)
-  {
-    reinterpret_cast<Gosu::Song*>( song )->stop();
-  }
-
-  double Gosu_Song_volume(Gosu_Song* song){
-    return reinterpret_cast<Gosu::Song*>( song )->volume();
-  }
-
-  void Gosu_Song_set_volume(Gosu_Song* song, double volume){
-    return reinterpret_cast<Gosu::Song*>( song )->set_volume(volume);
-  }
-
-  Gosu_Song* Gosu_Song_current_song()
-  {
+#include "Gosu_FFI_internal.h"
+
+GOSU_FFI_API Gosu_Song* Gosu_Song_create(const char* filename)
+{
+    return new Gosu_Song{filename};
+}
+
+GOSU_FFI_API void Gosu_Song_destroy(Gosu_Song* song)
+{
+    delete song;
+}
+
+GOSU_FFI_API void Gosu_Song_play(Gosu_Song* song, bool looping)
+{
+    song->play(looping);
+}
+
+GOSU_FFI_API bool Gosu_Song_playing(Gosu_Song* song)
+{
+    return song->playing();
+}
+
+GOSU_FFI_API void Gosu_Song_pause(Gosu_Song* song)
+{
+    song->pause();
+}
+
+GOSU_FFI_API bool Gosu_Song_paused(Gosu_Song* song)
+{
+    return song->paused();
+}
+
+GOSU_FFI_API void Gosu_Song_stop(Gosu_Song* song)
+{
+    song->stop();
+}
+
+GOSU_FFI_API double Gosu_Song_volume(Gosu_Song* song)
+{
+    return song->volume();
+}
+
+GOSU_FFI_API void Gosu_Song_set_volume(Gosu_Song* song, double volume)
+{
+    return song->set_volume(volume);
+}
+
+GOSU_FFI_API Gosu_Song* Gosu_Song_current_song(void)
+{
     return reinterpret_cast<Gosu_Song*>(Gosu::Song::current_song());
```

Comment:
This is the one reinterpret_cast that I couldn't get rid of. dynamic_cast doesn't work because Song is not polymorphic.

Probably not worth worrying about, because the whole concept of Gosu::Song::current_song will hopefully be obsolete sooner rather than later.